File: quick-sort.cpp
#include <stdio.h>
#include <chrono>

  void swap(int * array, size_t i, size_t store) {
      int temp = array[i];
      array[i] = array[store];
      array[store] = temp;
  }

  int median(int * array, size_t left, size_t right) {
        size_t mid = (left + right) / 2;
        if (array[right] < array[left]) {
              swap(array, left, right);
        }
        if (array[mid] < array[left]) {
              swap(array, mid, left);
        }
        if (array[right] < array[mid]) {
              swap(array, right, mid);
        }
        swap(array, right, mid);
        return array[right];
  }

  size_t mePartition(int * arr, size_t left, size_t right) {
        auto pivot = median(arr, left, right);
        auto leftIndex = left - 1;
        auto rightIndex = right;
        for (;;) {
            for (;;) {
                leftIndex += 1;
                if (!(arr[leftIndex] < pivot)) {
                    break;
                }
            }
            for (;;) {
                rightIndex -= 1;
                if (!(rightIndex > 0 && arr[rightIndex] > pivot)) {
                    break;
                }
            }

              if (leftIndex >= rightIndex) {
                    // indexes crossed
                    break;
              } else {
                    // just swapped
                    swap(arr, leftIndex, rightIndex);
              }
        }
        swap(arr, leftIndex, right);
        return leftIndex;
  }

  void quickSort(int * array, size_t left, size_t right) {
          if (left < right) {
                auto pivot = mePartition(array, left, right);
                quickSort(array, left, pivot - 1);
                quickSort(array, pivot + 1, right);
          }
  }

int main() {
    printf("C++\n");

    size_t len = 3000000;
    auto data = new int[len];

    for (int attempt = 0; attempt < 10; ++attempt) {
        for (int i = 0; i < len; ++i) {
            data[i] = -(int)i;
        }

        auto start = std::chrono::steady_clock::now();
        quickSort(data, 0, len-1);
        auto end = std::chrono::steady_clock::now();

        for (int i = 0; i < len - 1; ++i) {
            if (data[i] > data[i+1]) {
                printf("sort failed\n");
                break;
            }
        }

        printf("took %dms\n", (int)std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count());
    }
}

File: quick_sort.rs
use std::time::Instant;
fn main() {
    println!("Rust");

    let mut data = Box::new([0i32; 3000000]);

    for attempt in 0..10 {
        for i in 0..data.len() {
            data[i] = -(i as i32);
        }

        let start = Instant::now();
        let len = data.len();
        quickSort(&mut data[..], 0, len-1);
        let duration = start.elapsed();

        for i in 0..(data.len() - 1) {
            if data[i] > data[i+1] {
                eprintln!("sort failed\n");
                break;
            }
        }

        println!("took {}ms", duration.as_millis());
    }
}

  fn quickSort(array: &mut [i32], left: usize, right: usize) {
          if left < right {
                let pivot = mePartition(array, left, right);
                quickSort(array, left, pivot - 1);
                quickSort(array, pivot + 1, right);
          }
  }

  fn mePartition(arr: &mut [i32], left: usize, right: usize) -> usize {
        let pivot = median(arr, left, right);
        let mut leftIndex = left - 1;
        let mut rightIndex = right;
        loop {
            loop {
                leftIndex += 1;
                if !(arr[leftIndex] < pivot) {
                    break
                }
            }
            loop {
                rightIndex -= 1;
                if !(rightIndex > 0 && arr[rightIndex] > pivot) {
                    break
                }
            }

              if leftIndex >= rightIndex {
                    // indexes crossed
                    break;
              } else {
                    // just swapped
                    swap(arr, leftIndex, rightIndex);
              }
        }
        swap(arr, leftIndex, right);
        return leftIndex;
  }

  fn median(array: &mut [i32], left: usize, right: usize) -> i32 {
        let mid: usize = (left + right) / 2;
        if (array[right] < array[left]) {
              swap(array, left, right);
        }
        if (array[mid] < array[left]) {
              swap(array, mid, left);
        }
        if (array[right] < array[mid]) {
              swap(array, right, mid);
        }
        swap(array, right, mid);
        return array[right];
  }

  fn swap(array: &mut [i32], i: usize, store: usize) {
      let temp: i32 = array[i];
      array[i] = array[store];
      array[store] = temp;
  }
